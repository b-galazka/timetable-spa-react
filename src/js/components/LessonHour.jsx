import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import detectIE10 from '../utils/detectIE10';

class LessonHour extends Component {

    render() {

        const { mobileHidden, order, number } = this.props;
        const hasLessons = this.hasLessons();

        return (
            <article
                className={

                    classNames({
                        table__cell: true,
                        'table__cell--empty': (! hasLessons && ! mobileHidden),
                        'table__cell--mobile-hidden': mobileHidden
                    })
                }

                style={{ order }}
                data-number={number}
                ref={(el) => { this.tableCell = el; }}
            >
                {hasLessons && this.renderLessons()}
            </article>
        );
    }

    componentDidMount() {

        this.applyPrefix();
    }

    renderLessons() {

        const { lessons } = this.props;

        return lessons.map((lesson) => {

            if (! lesson) {

                return null;
            }

            const { _id, subject } = lesson;

            return (
                <div key={_id}>
                    <span>{subject}</span>
                    {this.renderSecondInfo(lesson)}
                    {this.renderThirdInfo(lesson)}
                </div>
            );
        });
    }

    renderSecondInfo(lesson) {

        const { timetableObjectType } = this.props;
        const { teacherSlug, class: schoolClass, teacherName } = lesson;

        switch (timetableObjectType) {

            case 'teacher':
                return <Link to={`/klasa/${encodeURIComponent(schoolClass)}`}>{schoolClass}</Link>;

            case 'class':
            case 'classroom':
                return (
                    <Link
                        to={`/nauczyciel/${encodeURIComponent(teacherSlug)}`}
                        title={teacherName}
                    >
                        {teacherSlug}
                    </Link>
                );

            default:
                return null;
        }
    }

    renderThirdInfo(lesson) {

        const { timetableObjectType } = this.props;
        const { classroom, class: schoolClass } = lesson;

        switch (timetableObjectType) {

            case 'classroom':
                return <Link to={`/klasa/${encodeURIComponent(schoolClass)}`}>{schoolClass}</Link>;

            case 'class':
            case 'teacher':
                return <Link to={`/sala/${encodeURIComponent(classroom)}`}>{classroom}</Link>;

            default:
                return null;
        }
    }

    hasLessons() {

        const { lessons } = this.props;

        return Array.isArray(lessons) && Boolean(lessons[0]);
    }

    applyPrefix() {

        if (detectIE10()) {

            const { style } = this.tableCell;

            style.msFlexOrder = style.order;
        }
    }
}

LessonHour.propTypes = {
    lessons: PropTypes.arrayOf(PropTypes.shape({
        _id: PropTypes.string.isRequired,
        subject: PropTypes.string.isRequired,
        class: PropTypes.string,
        classroom: PropTypes.string,
        teacherName: PropTypes.string,
        teacherSlug: PropTypes.string
    })),

    timetableObjectType: PropTypes.oneOf(['teacher', 'class', 'classroom']),
    mobileHidden: PropTypes.bool,
    order: PropTypes.number.isRequired,
    number: PropTypes.number
};

LessonHour.defaultProps = {
    lessons: null,
    timetableObjectType: null,
    mobileHidden: false,
    number: 0
};

export default LessonHour;