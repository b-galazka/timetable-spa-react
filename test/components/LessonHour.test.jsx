import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';

import LessonHour from '../../src/js/components/LessonHour';

Enzyme.configure({ adapter: new Adapter() });

describe('LessonHour component', () => {

    const { userAgent } = navigator;

    it('should render empty table cell', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <LessonHour order={1} />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render table cell hidden on mobile phones', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <LessonHour order={1} mobileHidden />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render teacher\'s lessons', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <LessonHour
                    timetableObjectType="teacher"
                    order={1}
                    number={1}
                    lessons={[

                        {
                            _id: "1",
                            subject: "math",
                            class: "3ET#",
                            classroom: "200"
                        },

                        {
                            _id: "2",
                            subject: "math",
                            class: "4ET",
                            classroom: "200"
                        }
                    ]}
                />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render class\'s lessons', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <LessonHour
                    timetableObjectType="class"
                    order={1}
                    number={1}
                    lessons={[

                        {
                            _id: "1",
                            subject: "math",
                            classroom: "200",
                            teacherSlug: 'XYZ',
                            teacherName: 'XYZ lastname'
                        },

                        {
                            _id: "2",
                            subject: "PE",
                            classroom: "300#",
                            teacherSlug: 'ABC',
                            teacherName: null
                        }
                    ]}
                />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render classroom\'s lessons', () => {

        const tree = renderer.create(
            <MemoryRouter>
                <LessonHour
                    timetableObjectType="class"
                    order={1}
                    number={1}
                    lessons={[

                        {
                            _id: "1",
                            subject: "math",
                            class: "3ET",
                            teacherSlug: 'XYZ',
                            teacherName: 'XYZ lastname'
                        },

                        {
                            _id: "2",
                            subject: "PE",
                            class: "4ET",
                            teacherSlug: '#ABC',
                            teacherName: null
                        }
                    ]}
                />
            </MemoryRouter>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should add style.msFlexOrder if app is running on IE10', () => {

        Object.defineProperty(navigator, 'userAgent', {
            value: 'lorem MSIE ipsum',
            configurable: true
        });

        const lessonHour = mount(
            <MemoryRouter>
                <LessonHour order={1} />
            </MemoryRouter>
        );

        const lessonHourStyles = lessonHour.find('.table__cell').getDOMNode().style;

        expect(lessonHourStyles.order).toBe(lessonHourStyles.msFlexOrder);
    });

    it('should add style.msFlexOrder if app is running on IE10 (regex)', () => {

        Object.defineProperty(navigator, 'userAgent', {
            value: 'lorem Tridentxrv:11. ipsum',
            configurable: true
        });

        const lessonHour = mount(
            <MemoryRouter>
                <LessonHour order={1} />
            </MemoryRouter>
        );

        const lessonHourStyles = lessonHour.find('.table__cell').getDOMNode().style;

        expect(lessonHourStyles.order).toBe(lessonHourStyles.msFlexOrder);   
    });

    afterEach(() => {

        Object.defineProperty(navigator, 'userAgent', {
            value: userAgent,
            configurable: true
        });
    });
});