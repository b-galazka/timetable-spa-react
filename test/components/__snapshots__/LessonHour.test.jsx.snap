// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonHour component should render class's lessons 1`] = `
<article
  className="table__cell"
  data-number={1}
  style={
    Object {
      "order": 1,
    }
  }
>
  <div>
    <span>
      math
    </span>
    <a
      href="/nauczyciel/XYZ"
      onClick={[Function]}
      title="XYZ lastname"
    >
      XYZ
    </a>
    <a
      href="/sala/200"
      onClick={[Function]}
    >
      200
    </a>
  </div>
  <div>
    <span>
      PE
    </span>
    <a
      href="/nauczyciel/ABC"
      onClick={[Function]}
      title={null}
    >
      ABC
    </a>
    <a
      href="/sala/300%23"
      onClick={[Function]}
    >
      300#
    </a>
  </div>
</article>
`;

exports[`LessonHour component should render classroom's lessons 1`] = `
<article
  className="table__cell"
  data-number={1}
  style={
    Object {
      "order": 1,
    }
  }
>
  <div>
    <span>
      math
    </span>
    <a
      href="/nauczyciel/XYZ"
      onClick={[Function]}
      title="XYZ lastname"
    >
      XYZ
    </a>
    <a
      href="/sala/undefined"
      onClick={[Function]}
    />
  </div>
  <div>
    <span>
      PE
    </span>
    <a
      href="/nauczyciel/%23ABC"
      onClick={[Function]}
      title={null}
    >
      #ABC
    </a>
    <a
      href="/sala/undefined"
      onClick={[Function]}
    />
  </div>
</article>
`;

exports[`LessonHour component should render empty table cell 1`] = `
<article
  className="table__cell table__cell--empty"
  data-number={0}
  style={
    Object {
      "order": 1,
    }
  }
/>
`;

exports[`LessonHour component should render table cell hidden on mobile phones 1`] = `
<article
  className="table__cell table__cell--mobile-hidden"
  data-number={0}
  style={
    Object {
      "order": 1,
    }
  }
/>
`;

exports[`LessonHour component should render teacher's lessons 1`] = `
<article
  className="table__cell"
  data-number={1}
  style={
    Object {
      "order": 1,
    }
  }
>
  <div>
    <span>
      math
    </span>
    <a
      href="/klasa/3ET%23"
      onClick={[Function]}
    >
      3ET#
    </a>
    <a
      href="/sala/200"
      onClick={[Function]}
    >
      200
    </a>
  </div>
  <div>
    <span>
      math
    </span>
    <a
      href="/klasa/4ET"
      onClick={[Function]}
    >
      4ET
    </a>
    <a
      href="/sala/200"
      onClick={[Function]}
    >
      200
    </a>
  </div>
</article>
`;
